text_generation/causal_lm/cpp/speculative_decoding.cpp

```diff
@@ -0,0 +1,300 @@
+}
+
+std::string detokenize(ov::InferRequest& detokenizer, std::vector<int64_t>& tokens) {
+    constexpr size_t BATCH_SIZE = 1;
+    detokenizer.set_input_tensor(ov::Tensor{ov::element::i64, {BATCH_SIZE, tokens.size()}, tokens.data()});
+    detokenizer.infer();
+    return detokenizer.get_output_tensor().data<std::string>()[0];
+}
+
+// The following reasons require TextStreamer to keep a cache of previous tokens:
+// detokenizer removes starting ' '. For example detokenize(tokenize(" a")) == "a",
+// but detokenize(tokenize("prefix a")) == "prefix a"
+// 1 printable token may consist of 2 token ids: detokenize(incomplete_token_idx) == "�"
+struct TextStreamer {
+    ov::InferRequest detokenizer;
+    std::vector<int64_t> token_cache;
+    size_t print_len = 0;
+
+    void put(int64_t token) {
+        token_cache.push_back(token);
+        std::string text = detokenize(detokenizer, token_cache);
+        if (!text.empty() && '\n' == text.back()) {
+            // Flush the cache after the new line symbol
+            std::cout << std::string_view{text.data() + print_len, text.size() - print_len};
+            token_cache.clear();
+            print_len = 0;
+	    return;
+        }
+        if (text.size() >= 3 && text.compare(text.size() - 3, 3, "�") == 0) {
+            // Don't print incomplete text
+            return;
+        }
+        std::cout << std::string_view{text.data() + print_len, text.size() - print_len} << std::flush;
+        print_len = text.size();
+    }
+
+    void end() {
+        std::string text = detokenize(detokenizer, token_cache);
+        std::cout << std::string_view{text.data() + print_len, text.size() - print_len} << '\n';
+        token_cache.clear();
+        print_len = 0;
+    }
+};
+}
+
+void forward_key_values(ov::InferRequest& request) {
+    // forwards present key/values from Results to Parameters with past key/values
+    // the first inputs are: input_ids, attention_masks, position_ids, other ones are past key/values.
+    // the first output is logits, others are present key.value
+    auto num_inputs = request.get_compiled_model().inputs().size();
+    for (size_t idx = 3; idx < num_inputs; ++idx) {
+        request.set_input_tensor(idx, request.get_output_tensor(idx - 2));
+    }
+}
+
+void init_key_values(ov::InferRequest request) {
+    auto num_inputs = request.get_compiled_model().inputs().size();
+    for (size_t idx = 3; idx < num_inputs; ++idx) {
+        auto kv_tensor = request.get_input_tensor(idx);
+        ov::Shape tensor_shape = kv_tensor.get_shape();
+        kv_tensor.set_shape({BATCH_SIZE, tensor_shape[1], 0, tensor_shape[3]});
+    }
+}
+
+ov::Tensor trimm_tensor(ov::Tensor tensor, uint64_t new_seq_len) {
+    // copy elements from the old to the new tensor
+    // tensor shape = [batch, size_1, seq_len, size_2]
+    auto old_tensor_data = tensor.data<float>();
+    auto shape = tensor.get_shape();
+    size_t size_1 = shape[1];
+    size_t old_seq_len = shape[2];
```

Comment:
let's also try to support cases when seq_len is first dimension

Reply:
Done. I was unable to verify on chatglm since it failed on intel-optimum during conversion. But it should work

Reply:
you can use llm_bench to convert the model